Ph sync fix

patrickhill opened the pull request:
Fixes several bugs. Makes sync work again.

patrickhill committed:
fixes depricated removeObject method by replacing it with removeSharedObjectAtIndex method

patrickhill committed:
replaces depricated setLineSpacing method with setLineHeight method

patrickhill committed:
comments out setCharacterSpancing method because it is breaking the plugin

patrickhill committed:
removes commented out lines

patrickhill committed:
fixes setCharacterSpacing bug by first converting style.Character to a number

patrickhill committed:
clears existing type styles before importing because sketch will freeze if the to be imported styles already exist

patrickhill committed:
restores broken color swatch import with updated applyColors object variable and addColors method

patrickhill committed:
comments out removeAllSharedObjects line because it's dangerous. Could cause someone to loose all there document styles

nolastan commented on Contents/Sketch/colors.js:
```diff
@@ -7,7 +7,5 @@ var applyColors = function (newColors, docData) {
   }
-  colors = MSArray.dataArrayWithArray(colors);
-  appController.globalAssets().setPrimitiveColors(colors);
-  appController.globalAssets().objectDidChange();
```
Thanks so much for working on this! `objectDidChange` is necessary for the inspector pane to update if a text layer is already selected. Is there a reason for removing this?

patrickhill replied:
Nope, I was literally just hacking away to make it work. I'm making mistakes like this because I'm a noob.

nolastan replied:
Cool no problem, let's just put it back then.

nolastan commented on Contents/Sketch/styles.js:
```diff
@@ -43,8 +43,7 @@ var applyStyles = function (newStyles, sharedStyles) {
     while(existingStyles.count() > 0) {
-      style = existingStyles.objectAtIndex(0);
-      [existingStyles removeObject:style];
+      [sharedStyles removeSharedObjectAtIndex:0];
```
Layer Styles doesn't seem to work at all.

![image](https://cloud.githubusercontent.com/assets/727778/17226978/f172d758-54c0-11e6-8a6d-bf349dd4eed8.png)

nolastan replied:
Ah, looking below this function is no longer called.

patrickhill replied:
Yup, the color swatches didn't work either. I got the typography working then the colors. Haven't made it to the layer styles yet.

nolastan commented on Contents/Sketch/typography.js:
```diff
@@ -7,7 +7,10 @@ var applyTypography = function (newStyles, sharedStyles) {
     'justified': 3
   };
 
-  removeAllStyles();
```
If we're not going to call this, we should just remove the function entirely. And also update the README to indicate that styles will be duplicated every time the plugin is run.

patrickhill replied:
I can't decide what to do on this one. Neither of the solutions seem ideal. I might just leave it the way you had it.

I'm almost finished with a new feature that will:
1. look for any existing text styles that match style names in the spreadsheet
2. then update them with any changes in the spreadsheet
3. then append any additional styles

nolastan replied:
If you could get that working it'd be amazing. I've spent many many hours trying!

patrickhill replied:
I think I figured out. I'll be pushing a new branch for you to test out in a few minutes.

nolastan commented on Contents/Sketch/typography.js:
```diff
@@ -19,8 +22,11 @@ var applyTypography = function (newStyles, sharedStyles) {
     if("Size"      in style)  { textLayer.setFontSize(style.Size); }
-    if("Line"      in style)  { textLayer.setLineSpacing(style.Line); }
-    if("Character" in style)  { textLayer.setCharacterSpacing(style.Character); }
+    if("Line"      in style)  { textLayer.setLineHeight(style.Line); }
```
Nice, they finally updated this deprecation message! It used to say `Warning: -[MSTextLayer setLineSpacing:] is deprecated. Please update your script to use -[MSTextLayer setLineSpacing:] instead.` so I had no idea what to do 😄

nolastan commented on Contents/Sketch/typography.js:
```diff
@@ -19,8 +22,11 @@ var applyTypography = function (newStyles, sharedStyles) {
+    if("Character" in style)  { 
+      var characterSpacing = Number(style.Character);
+      textLayer.setCharacterSpacing(characterSpacing); 
+    }
```
Cool, this is better than my fix (which I had been begging people to test before merging in) https://github.com/nolastan/sync.sketchplugin/compare/character-fix

patrickhill committed:
compares names of shared text styles and only imports new ones

patrickhill committed:
brings back the remove existing styles functionality

patrickhill committed:
brings back accidently deleted objectDidChange() method

patrickhill committed:
only imports new styles without duplication

patrickhill committed:
if an imported text style has the same name as an existing one, the existing style will now be updated with the settings from the new one. This allows you to update all text elements in the document that use that shared text style

patrickhill committed:
removes console logs and unused commented out code

patrickhill committed:
updates deprecated addNewStylePart method

patrickhill committed:
merge

nolastan commented on Contents/Sketch/styles.js:
```diff
@@ -1,5 +1,14 @@
+  // layer.style().addStylePartOfType(1) // To add a new border
+  // layer.style().addStylePartOfType(2) // To add a new shadow
+  // layer.style().addStylePartOfType(3) // To add a new inner shadow
+
```
Nice find. Perhaps we enumerate [like this](https://github.com/nolastan/sync.sketchplugin/blob/master/Contents/Sketch/typography.js#L3) so we don't need all the commenting. (There might be a more formal way to enumerate in cocoascript/objective-c). Not a blocker but just a suggestion for code clarity.

nolastan commented on Contents/Sketch/typography.js:
```diff
@@ -8,12 +10,41 @@ var applyTypography = function (newStyles, sharedStyles) {
+      if(existingName == newStyleName) {
+        // log("they match");
+
+        sharedStyles.updateValueOfSharedObject_byCopyingInstance(existingStyleObjects[i], newStyle);
```
Awesome! Will verify this works shortly. Let's make sure not to commit commented out log messages.

nolastan replied:
Also, curious, how did you figure this out?

patrickhill replied:
Cool, I'll clean it up.

I used this: 
https://github.com/sketchplugins/plugin-requests/blob/master/Dump%20Selected%20Object.sketchplugin
to log out the sharedStyles object. Then I went through all the methods looking for something that might work. It was basically just trial and error. I went through several before I found updateValueOfSharedObject_byCopyingInstance.

patrickhill committed:
cleans up code be removing log comments and unneeded objectID return on checkForMatchingStyleAndMerge

patrickhill committed:
removes unnecessary return false on checkForMatchingStyleAndMerge

patrickhill committed:
comments out objectDidChange() because it was throwing error: not a function

patrickhill committed:
adds alignmentHash for code clarity

patrickhill committed:
adds comment to error line

patrickhill commented on Contents/Sketch/colors.js:
```diff
@@ -7,7 +7,9 @@ var applyColors = function (newColors, docData) {
+  // The following line is throwing the error: is not a function
+  // appController.globalAssets().objectDidChange();
+  
 }
```
@nolastan objectDidChange() line is throwing an error so I re-commented it out to finish up the other changes. My first guess is the method is deprecated. I'm taking a break on this repo for now. Can you look into a fix if you want the line back in there?

nolastan commented:
Testing this in sketch with the default Google Sheet, it works fine on first run but upon running the plugin a second time Sketch freezes up. I haven't dug in further but I suspect the update check is looping infinitely.

patrickhill committed:
fixes freeze when sync is run a second time be replacing removeAllStyles function in styles.js with the removeAllSharedObjects method

patrickhill commented:
@nolastan I believe I fixed this issue. It looks like it was the removeAllStyles function. I replaced with a built in method that does the same thing. I've pushed the change for you to test.

nolastan commented:
Great – it's working now! Document colors duplicate on each run but we can deal with that separately.

patrickhill commented:
Sounds good!
